### ISSUE TITLE
`bootstrap.js` doesn't work with ESM

### ISSUE THREAD

**Issue by gustavopch** [firebase/firebase-tools#5847]
The emulator starts but then crashes with the following error when I make a request to my Express-based website:
```
ReferenceError: exports is not defined in ES module scope
>  This file is being treated as an ES module because it has a '.js' file extension and '<path-to-project)/.firebase/<site-id>/functions/package.json' contains "type": "module". To treat it as a CommonJS script, rename it to use the '.cjs' file extension.
```

That's because it's using `exports.handle`. It should be `export const handle` when the application is using ESM.

https://github.com/firebase/firebase-tools/blob/56fb8a34201ab903f92178f696bc3b28acfb79bb/src/frameworks/express/index.ts#L64-L69

**Comment by gustavopch** [comment]
@jamesdaniels I may also have found another issue (or I just don't know how to use ESM, which is also pretty much possible). When the project uses ESM, `dynamicImport` is called with an absolute path to the project directory and throws this error:
```
Error [ERR_UNSUPPORTED_DIR_IMPORT]: Directory import '/Users/gustavopch/Projects/my-project' is not supported resolving ES modules imported from /Users/gustavopch/Projects/my-project/node_modules/firebase-tools/lib/dynamicImport.js
```
https://github.com/firebase/firebase-tools/blob/56fb8a34201ab903f92178f696bc3b28acfb79bb/src/frameworks/express/index.ts#L57-L60

It works when I apply this patch, but of course it only works if the entry file is called `index.js`:
```diff
-      entry = await dynamicImport(root).catch(() => undefined);
+      entry = await dynamicImport(`${root}/index.js`).catch(() => undefined);
```

It was a bit hard to trace it down because it's being swallowed in the catch callback.

**Comment by gustavopch** [comment]
Yeah, seems that's really how `import` works. If you provide a path instead of a package name, you must point to the exact file, not to the directory. So the solution will probably be something like:
```diff
-      entry = await dynamicImport(root).catch(() => undefined);
+      entry = await dynamicImport(join(root, packageJson.module || packageJson.main));
```
I removed the `catch(() => undefined)` because it's better to crash the whole emulator with a helpful message than just silently not work.

**Comment by erdebaru** [comment]
any news on this? 
encountered same issue

**Comment by gustavopch** [comment]
You'll have to patch the package yourself while they don't fix it.

In my case, I decided to go with my own implementation (studied the source code of `firebase-tools` and `firebase-frameworks` to do that) instead of using the built-in web frameworks feature, so I don't have this (or any other) problem anymore and my implementation works with Remix's HMR. It also doesn't require the codegen nor the `npm install` steps, making it much simpler and faster to run in the emulator.

Side note: I don't think I get why they're following this complicated architecture to implemented web frameworks support; seems over-engineered to me, but I'm probably missing something.

**Comment by pascalconfluent** [comment]
Good morning!

Any update on this issue. I do have the same issue :(
Would be super nice to have an official fix!

Thank you!

**Comment by pascalconfluent** [comment]
By the way it is not only an emulator issue, it happens with the latest of firebase. 
Changing the package type to commonjs instead of module seems to work.

**Comment by pascalconfluent** [comment]
@gustavopch would you be ok to share more about your solution?

**Comment by gustavopch** [comment]
@pascalconfluent I'm using Remix. I have this `ssr` function:

```ts
import { onRequest } from 'firebase-functions/v2/https'

import { assert } from '../../app/misc.js'

// Firebase Emulator loads this module twice: first simply for discovery, then
// to actually handle requests. In the second case, we need to load the server
// module because it will tell Remix the server is ready. In the first case
// though, we shouldn't load the server module because the required environment
// variables are not loaded and it would crash.
//
// IMPORTANT: after HMR, Firebase will kill the worker that runs this module
// and only restart it when a new request arrives, rendering HMR useless as such
// request won't happen. As a workaround, we use '--inspect-functions', which
// changes the behavior so the worker is immediately restarted when the code is
// changed.
const server = process.env.FUNCTIONS_CONTROL_API
  ? null
  : await import('../../server.js')

export const ssr = onRequest(
  { memory: '2GiB', timeoutSeconds: 540 },
  (request, response) => {
    assert(server, 'server.js not loaded')

    server.app(request, response)
  },
)
```

I simply pass the `request` and `response` given by Firebase to the Express instance (i.e. `server.app`). And then my `firebase.json` looks like this:
```json
{
  "functions": {
    "runtime": "nodejs18",
    "source": "."
  },
  "hosting": {
    "public": "./build/client",
    "headers": <omitted for brevity>,
    "cleanUrls": true,
    "rewrites": [
      {
        "source": "**",
        "function": {
          "functionId": "ssr",
          "pinTag": true
        }
      }
    ]
  }
}
```

Simple as that.

**Comment by akashlama1998-icloud** [comment]
This hasn't fixed yet.. because of it I have to set remix up with cloud function manually which is fine.. but web frameworks gives us the current logged in user and app in res.locals which manually deploying doesn't.. I really needed that feature but instead I have to manage logged in user server side by myself 😔. I hope the developer notice this bug and it get's fixed or they officially support remix like every other framework and we wouldn't have to fiddle around with Integrate other frameworks with Express.js solution, anyone will work.. but please fix this issue soon.
